client/allochealth/tracker.go

```diff
@@ -277,15 +277,35 @@ func (t *Tracker) watchTaskEvents() {
 		// Detect if the alloc is unhealthy or if all tasks have started yet
 		latestStartTime := time.Time{}
 		for taskName, state := range alloc.TaskStates {
-			// If the task is a poststop task we do not want to evaluate it
-			// since it will remain pending until the main task has finished
-			// or exited.
-			if t.lifecycleTasks[taskName] == structs.TaskLifecycleHookPoststop {
-				continue
+			if t.lifecycleTasks[taskName] != nil {
+				// If the task is a poststop task we do not want to evaluate it
+				// since it will remain pending until the main task has finished or
+				// exited.
+				if t.lifecycleTasks[taskName].Hook == structs.TaskLifecycleHookPoststop {
+					continue
+				}
+
+				// If this is a poststart task which has already succeeded we
+				// want to check for two possible success conditions before
+				// attempting to evaluate it.
+				if t.lifecycleTasks[taskName].Hook == structs.TaskLifecycleHookPoststart && state.Successful() {
+
+					// If the task was successful and it's runtime is at least
+					// t.minHealthyTime, skip evaluation.
+					if state.FinishedAt.Sub(state.StartedAt) >= t.minHealthyTime {
+						continue
+					}
+
+					// If the task was successful and the user set
+					// 'ignore_min_healthy_time' to 'true', skip evaluation.
+					if t.lifecycleTasks[taskName].IgnoreMinHealthyTime {
+						continue
+					}
+				}
 			}
 
 			// One of the tasks has failed so we can exit watching
-			if state.Failed || (!state.FinishedAt.IsZero() && t.lifecycleTasks[taskName] != structs.TaskLifecycleHookPrestart) {
+			if state.Failed || (!state.FinishedAt.IsZero() && t.lifecycleTasks[taskName].Hook != structs.TaskLifecycleHookPrestart) {
```

Comment:
Hi @beautifulentropy! We're going to take some time to review this in more depth but right now a bunch of tests are failing with a nil pointer dereference here:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1399289]

goroutine 134 [running]:
github.com/hashicorp/nomad/client/allochealth.(*Tracker).watchTaskEvents(0xc0007c1a40)
	/home/circleci/go/src/github.com/hashicorp/nomad/client/allochealth/tracker.go:308 +0xe49
created by github.com/hashicorp/nomad/client/allochealth.(*Tracker).Start
	/home/circleci/go/src/github.com/hashicorp/nomad/client/allochealth/tracker.go:139 +0x5f
```

I think that's because the `t.lifecycleTasks` might not have a key for `taskName`, which was safe before this change because the value of `t.lifecycleTasks["nonexistenttask"]` would just be an empty string we could compare. But now it's a pointer and dereferencing it to get the `.Hook` field panics if it's nil.

Reply:
Ah, thanks! I've gone ahead and addressed this.